File: rust/notify-agent-core/src/toast_xml.rs
use std::path::Path;

use crate::model::ImageShape;
use crate::toast::ToastRequest;

pub fn xml_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&apos;")
}

/// file:/// URI with forward slashes and percent-encoded segments; a leading
/// '/' (unix paths) is not doubled. Tuned for the two path shapes this
/// codebase produces (unix absolute, windows drive-letter), not general URIs.
fn file_uri(path: &Path) -> String {
    let p = path.display().to_string().replace('\\', "/");
    let p = p.strip_prefix('/').unwrap_or(&p);
    let mut encoded = String::with_capacity(p.len());
    for b in p.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9'
            | b'-' | b'.' | b'_' | b'~' | b'/' | b':' => encoded.push(b as char),
            _ => encoded.push_str(&format!("%{b:02X}")),
        }
    }
    format!("file:///{encoded}")
}

/// Toast XML per design §7 budget: ≤3 texts, 1 button, plus the separate
/// appLogoOverride image slot (design 2026-07-22). image_path is the LOCAL
/// cached file — the OS ignores remote URLs for unpackaged apps, so a failed
/// fetch (None) renders imageless.
pub fn build_toast_xml(toast: &ToastRequest, image_path: Option<&Path>) -> String {
    let image = match (image_path, &toast.image) {
        (Some(path), Some(image_ref)) => {
            let crop = match image_ref.shape {
                ImageShape::Circle => r#" hint-crop="circle""#,
                ImageShape::Square => "",
            };
            format!(
                r#"<image placement="appLogoOverride"{crop} src="{}"/>"#,
                xml_escape(&file_uri(path))
            )
        }
        _ => String::new(),
    };
    let attribution = toast
        .attribution
        .as_deref()
        .map(|a| format!(r#"<text placement="attribution">{}</text>"#, xml_escape(a)))
        .unwrap_or_default();
    let actions = match (&toast.action_label, &toast.action_url) {
        (Some(label), Some(url)) => format!(
            r#"<actions><action content="{}" arguments="{}" activationType="foreground"/></actions>"#,
            xml_escape(label),
            xml_escape(url)
        ),
        _ => String::new(),
    };
    format!(
        r#"<toast><visual><binding template="ToastGeneric">{image}<text>{}</text><text>{}</text>{attribution}</binding></visual>{actions}</toast>"#,
        xml_escape(&toast.title),
        xml_escape(&toast.message)
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{ImageRef, ImageShape};
    use crate::toast::ToastRequest;

    fn toast(image: Option<ImageRef>) -> ToastRequest {
        ToastRequest {
            title: "Tony Redmond".into(),
            message: "is now available".into(),
            attribution: Some("Microsoft Teams".into()),
            action_label: Some("Open <chat>".into()),
            action_url: Some("https://teams.example/chat?a=1&b=2".into()),
            sources: Vec::new(),
            image,
        }
    }

    #[test]
    fn no_image_builds_todays_xml() {
        let xml = build_toast_xml(&toast(None), None);
        assert!(xml.starts_with("<toast><visual><binding template=\"ToastGeneric\">"));
        assert!(xml.contains("<text>Tony Redmond</text>"));
        assert!(!xml.contains("<image"));
        // escaping in action attributes
        assert!(xml.contains("Open &lt;chat&gt;"));
        assert!(xml.contains("a=1&amp;b=2"));
    }

    #[test]
    fn circle_image_gets_applogo_with_crop() {
        let image = ImageRef { url: "https://x/a.jpg".into(), shape: ImageShape::Circle };
        let xml = build_toast_xml(&toast(Some(image)), Some(std::path::Path::new("/tmp/cache/abc123")));
        assert!(xml.contains(r#"<image placement="appLogoOverride" hint-crop="circle" src="file:///tmp/cache/abc123"/>"#));
    }

    #[test]
    fn square_image_omits_crop_attribute() {
        let image = ImageRef { url: "https://x/a.jpg".into(), shape: ImageShape::Square };
        let xml = build_toast_xml(&toast(Some(image)), Some(std::path::Path::new("/tmp/cache/abc123")));
        assert!(xml.contains(r#"<image placement="appLogoOverride" src="file:///tmp/cache/abc123"/>"#));
        assert!(!xml.contains("hint-crop"));
    }

    #[test]
    fn image_ref_without_local_path_renders_imageless() {
        let image = ImageRef { url: "https://x/a.jpg".into(), shape: ImageShape::Circle };
        let xml = build_toast_xml(&toast(Some(image)), None); // fetch failed
        assert!(!xml.contains("<image"));
    }

    #[test]
    fn windows_backslash_paths_become_forward_slash_file_uris() {
        let image = ImageRef { url: "https://x/a.jpg".into(), shape: ImageShape::Square };
        let xml = build_toast_xml(
            &toast(Some(image)),
            Some(std::path::Path::new(r"C:\Users\u\AppData\Local\DesktopNotificationAgent\image-cache\abc")),
        );
        assert!(xml.contains(r#"src="file:///C:/Users/u/AppData/Local/DesktopNotificationAgent/image-cache/abc"/>"#));
    }

    #[test]
    fn paths_with_spaces_are_percent_encoded() {
        let image = ImageRef { url: "https://x/a.jpg".into(), shape: ImageShape::Square };
        let xml = build_toast_xml(
            &toast(Some(image)),
            Some(std::path::Path::new(r"C:\Users\John Smith\AppData\Local\DesktopNotificationAgent\image-cache\abc")),
        );
        assert!(xml.contains(r#"src="file:///C:/Users/John%20Smith/AppData/Local/DesktopNotificationAgent/image-cache/abc"/>"#));
    }
}

File: golang/internal/windowstoast/toastxml.go
// Package windowstoast builds the toast notification XML consumed by
// Windows.UI.Notifications.ToastNotification, mirroring
// rust/notify-agent-core/src/toast_xml.rs byte-for-byte in shape. It is a
// pure, cross-platform-testable string builder: the only Windows-specific
// part of rendering — invoking PowerShell to actually submit the XML — lives
// in cmd/notify-agent-windows, not here.
package windowstoast

import (
	"fmt"
	"strings"

	"github.com/YuShuanHsieh/os-notification/golang/internal/httpsurl"
	"github.com/YuShuanHsieh/os-notification/golang/internal/model"
	"github.com/YuShuanHsieh/os-notification/golang/internal/toast"
)

var xmlEscaper = strings.NewReplacer(
	"&", "&amp;",
	"<", "&lt;",
	">", "&gt;",
	`"`, "&quot;",
	"'", "&apos;",
)

// XMLEscape escapes the five reserved XML characters in s.
func XMLEscape(s string) string {
	return xmlEscaper.Replace(s)
}

// fileURI builds a file:/// URI from a local path with forward slashes and
// percent-encoded segments; a leading '/' (unix paths) is not doubled. This
// mirrors toast_xml.rs's file_uri, tuned for the two path shapes this
// codebase produces (unix absolute, Windows drive-letter), not general URIs.
func fileURI(path string) string {
	p := strings.ReplaceAll(path, `\`, "/")
	p = strings.TrimPrefix(p, "/")

	var b strings.Builder
	b.WriteString("file:///")
	for i := 0; i < len(p); i++ {
		c := p[i]
		switch {
		case c >= 'A' && c <= 'Z', c >= 'a' && c <= 'z', c >= '0' && c <= '9',
			c == '-', c == '.', c == '_', c == '~', c == '/', c == ':':
			b.WriteByte(c)
		default:
			fmt.Fprintf(&b, "%%%02X", c)
		}
	}
	return b.String()
}

// BuildToastXML builds the toast XML for req. cachedImagePath is the local
// on-disk path of req.Image, already resolved by the image cache — pass ""
// when there is no image or the fetch failed (the OS ignores remote URLs for
// unpackaged apps, so a failed fetch renders imageless rather than erroring).
//
// Shape: <toast><visual><binding template="ToastGeneric">{image}<text>title</text>
// <text>message</text>{attribution}</binding></visual>{actions}</toast>, where
// {image} is present only when both req.Image and cachedImagePath are set,
// {attribution} is present only when req.Attribution is non-empty, and
// {actions} is present only when both ActionLabel and ActionURL are
// non-empty AND ActionURL passes httpsurl.Valid — an invalid or partial
// action never fails the toast, it just omits the button.
func BuildToastXML(req toast.ToastRequest, cachedImagePath string) string {
	image := ""
	if cachedImagePath != "" && req.Image != nil {
		crop := ""
		if req.Image.Shape == model.ImageShapeCircle {
			crop = ` hint-crop="circle"`
		}
		image = fmt.Sprintf(`<image placement="appLogoOverride"%s src="%s"/>`,
			crop, XMLEscape(fileURI(cachedImagePath)))
	}

	attribution := ""
	if req.Attribution != "" {
		attribution = fmt.Sprintf(`<text placement="attribution">%s</text>`, XMLEscape(req.Attribution))
	}

	actions := ""
	if req.ActionLabel != "" && req.ActionURL != "" && httpsurl.Valid(req.ActionURL) {
		actions = fmt.Sprintf(
			`<actions><action content="%s" arguments="%s" activationType="protocol"/></actions>`,
			XMLEscape(req.ActionLabel), XMLEscape(req.ActionURL),
		)
	}

	return fmt.Sprintf(
		`<toast><visual><binding template="ToastGeneric">%s<text>%s</text><text>%s</text>%s</binding></visual>%s</toast>`,
		image, XMLEscape(req.Title), XMLEscape(req.Message), attribution, actions,
	)
}
